### DISCUSSION TITLE
Structure and organisation of functions files

### DISCUSSION THREAD

**Discussion by Tom687** [parcel-bundler/parcel#5899] [Q&A]
Hi, I would like to have some insight about the way I structure the functions in the files in bundle with Parcel.

Little bit of background : I am developing a web app using Node on the backend, PUG / JADE for the HTML side and vanilla JS for the client side JS.

The project includes lots of JS functions, that I put in dedicated files corresponding to the view files (ex : all the client side JS for files.pug is in files.js). I also use helper JS files when some functions are repeated. I then bundle all the client side JS files with Parcel.

The thing is, I am not sure about the way I handle all of this. If possible, it would to know what should be considered best practice solutions for the problems I am facing.

Let's get to the subject :

<br>
As an example, I have a function ( searchUser() ) used to display search results in a box below text inputs. When a user types in a search input, an AJAX call is made to the backend to fetch the DB. The results are then returned to the frontend. Then the function creates the result box that is populated with the results.

For this function to work, it uses functions from 2 files : searchHelpers.js and files.js. The function itself is in searchHelpers.js

Here are the used functions from searchHelpers.js for this to work **(as I feel the code of the functions itself is not relevant, I will just include the parts I am talking about in my questions)** :

```
// "Main" function, that is called from index.js
export const searchUser = (event) => {
    //… Code for the function to work
	
     // Relevant part : chained functions called from the same file (searchHelpers.js) :
    search(event.target.value.trim())
        .then(searchResults => {
	    populateSearchBox(searchResults);
        });
}


// "Internal" function, not exported, used only locally in searchResults.js
async function search(input) {
    //... Code making a GET request to backend
 
    if (res.status === 200 || res.data.status === 'success') {
        return res.data.searchResults; // Relevant part : function returns search results
    }
}


// "Internal" function, not exported, used only locally in searchResults.js
function populateSearchBox(searchResults) {
    //… Code in order for the function to work … Loop through the search results fetched from search() function above
    
    // Relevant part : Use a function, imported from another file, depending on a condition (if request comes from files.pug or goals.pug)
    if (searchInput.classList.contains('js-search-user-file')) {
        addFileLinkBtns(); // Relevant : Imported from "files.js"
    }
    else if (searchInput.classList.contains('js-search-user-goals')) {
        addGoalChecks(); // Relevant : Imported from "goals.js"
   }
}
```

Here is the corresponding part in index.js :

```
const searchUserBar = document.getElementById('searchUser');

if (searchUserBar) {
    // When typing in search bar, get search results from DB and insert it in search box
    searchUserBar.addEventListener('input', event => {
	searchUser(event);
    });
}
```

The work flow is like this :

index.js : calls the searchUsers() function when it catches a user typing in a input#searchUser.

The searchUser() function is imported in index.js from searchHelpers.js, which then calls search() and then populateSearchBox() (those 2 functions are from the same file, not exported functions).

From there, populateSearchBox() will have conditions to check what it should do next. Let's assume it should populate the search box from files.pug : it will call the addFileLinkBtns() function, that is exported from files.js and imported in searchHelpers.js. This function itself is not relevant, so I will not include it.

So this is the basic workflow I got right now, and I use it for more than a few functions.

<br>

**I am looking for better maintainability and legible code, as well as best practice solutions.**

What I would like to know is :

_**1) Is it OK to have functions exported and imported in different files, without being used in index.js ?**_ 
Just like the workflow I shared above : the "main" function ( searchUser() ) is imported in index.js from searchHelpers.js, which itself imports functions from files.js and goals.js, which themselves can import functions from other places.

_**2) Is it OK to have "internal" functions, not exported anywhere, used by other functions from the same .js file they are into ?**_ (Just like populateSearchBox() )

_**3) Is it OK to have functions calling functions calling functions ?**_ Just like the workflow above.

_**4) As I am trying to understand the best practices to use Parcel, could you point me to some directions to improve my code / structure / workflow ?**_

**I am using Parcel like this because I could not find any good real world example of its use, beside "Hello world" apps. If you have some good real world example of its use, please let me know**.

<br>

**Please let me know if I got it totally wrong, because I kinda fell that I am going a bit freestyle with this**.

Thank you very much for your time and your help

**Comment by mischnic** [comment]
Questions 1-3: Yes, that's very common. From a technical perspective, it's fine as long as it builds/runs without an error. So it's really up to you how you find it easiest to work with.

Another pattern is also to have an index file for example at `src/utils/index.js` that reexports helpers from various individual files: https://github.com/parcel-bundler/parcel/blob/v2/packages/core/utils/src/index.js. So you can always do `import {something} from "./utils/"` and don't have to repeat the name while still having the codebase split up into multiple smaller files.

4: Not sure, most guides about imports/exports are about how the syntax works, but you seem to mostly know that already.

**Reply by Tom687** [reply]
Hi, thank you very much for your answer.

About the 'export everything' pattern you are talking about, I don't really get the point. Is it to be able to import our functions from the same location and not from different files ? 
Also, by "you don't have to repeat the name", do you mean that, when importing, I don't have to specify the filename but rather just the directory it comes from (in this case, "/utils" ? Is that it or is there more to what you meant ?

Lastly, as I have yet not found any real world code applications with Parcel and vanilla JS, would you say the way I use it is correct and viable for a real app ? Is it "the Parcel way" of doing things ? Do you think of any kind of rule that I should follow ?

Thank you very much, I greatly appreciate your help

**Reply by mischnic** [reply]
> Is that it or is there more to what you meant ?

Yes, that's it.

> Is it "the Parcel way" of doing things

This isn't specific to Parcel, it's Javascript generally. Node, Webpack, Rollup, ... all behave the same here. If there are rules at all, they are usually conventions per project.

**Reply by Tom687** [reply]
Thank you very much for your help !
